src/Passes/Source/ProfilePass/Profile.cpp

```diff
@@ -0,0 +1,680 @@
+
+                // Finding argument constants
+                while (idx < n)
+                {
+                    auto arg   = callee_function->getArg(idx);
+                    auto value = call_instr->getArgOperand(idx);
+
+                    auto cst = llvm::dyn_cast<llvm::ConstantInt>(value);
+                    if (cst != nullptr)
+                    {
+                        argument_constants[arg->getName().str()] = cst;
+                    }
+                    else
+                    {
+                        remaining_arguments.push_back(idx);
+                    }
+
+                    ++idx;
+                }
+
+                // Making a function copy
+                auto new_callee = expandFunctionCall(*callee_function, argument_constants);
+
+                // Replacing call if a new function was created
+                if (new_callee != nullptr)
+                {
+
+                    llvm::IRBuilder<> builder(call_instr);
+
+                    // List with new call arguments
+                    std::vector<llvm::Value*> new_arguments;
+                    for (auto const& i : remaining_arguments)
+                    {
+                        // Getting the i'th argument
+                        llvm::Value* arg = call_instr->getArgOperand(i);
+
+                        // Adding arguments that were not constant
+                        if (argument_constants.find(arg->getName().str()) == argument_constants.end())
+                        {
+                            new_arguments.push_back(arg);
+                        }
+                    }
+
+                    // Creating a new call
+                    auto* new_call = builder.CreateCall(new_callee, new_arguments);
+                    new_call->takeName(call_instr);
+
+                    // Replace all calls to old function with calls to new function
+                    instr.replaceAllUsesWith(new_call);
+
+                    // Deleting instruction
+                    schedule_instruction_deletion.push_back(&instr);
+                    // TODO(tfr): Delete instruction, instr.deleteInstru??;
+
+                    constantFoldFunction(*new_callee);
+
+                    // Recursion: Returning the new call as the instruction to be analysed
+                    ret = new_call;
+                }
+
+                // Deleting the function the original function if it is no longer in use
+                if (callee_function->use_empty())
+                {
+                    callee_function->eraseFromParent();
+                }
+            }
+        }
+
+        return ret;
+    }
+
+    llvm::Value* ProfilePass::detectActiveCode(llvm::Value* input, DeletableInstructions&)
+    {
+        active_pieces_.insert(input);
+        return input;
+    }
+
+    bool ProfilePass::runOnFunction(llvm::Function& function, InstructionModifier const& modifier)
+    {
+        if (depth_ >= max_recursion_)
```

Comment:
How is the max_recursion_ specified? Is this a hardcoded value somewhere? How is that value chosen?

Reply:
This is a constructor paramater of the pass class. Default is 512 and it will be a pass configuration parameter in time.